File: sorting algorithms/merge sort/merge_sort.py
def merge_sort(array): #splitting
    if len(array) <= 1:
        return array
    mid = len(array) // 2
    left = array[:mid]
    right = array[mid:]

    left = merge_sort(left)
    right = merge_sort(right)

    return merge(left, right)

def merge(left, right):
    l = 0
    r = 0
    sorted = []
    while l < len(left) and r < len(right):
        if left[l] < right[r]:
            sorted.append(left[l])
            l += 1
        else:
            sorted.append(right[r])
            r += 1
        
    while l < len(left):
        sorted.append(left[l])
        l += 1
    while r < len(right):
        sorted.append(right[r])
        r += 1

    return sorted


array = [5, 4, 2, 3, 1, 7, 9]
print(merge_sort(array))

File: sorting algorithms/merge sort/merge_sort.js
function mergeSort(arr) {
    if (arr.length <= 1) {
        return arr;
    }

    const mid = Math.floor(arr.length / 2);
    const left = arr.slice(0, mid);
    const right = arr.slice(mid);

    return merge(mergeSort(left), mergeSort(right));
}

function merge(left, right) {
    let sortedArray = [];
    let l = 0, r = 0;

    while (l < left.length && r < right.length) {
        if (left[l] < right[r]) {
            sortedArray.push(left[l]);
            l++;
        } else {
            sortedArray.push(right[r]);
            r++;
        }
    }

    return sortedArray.concat(left.slice(l)).concat(right.slice(r));
}

// Example usage:
const arr = [12, 11, 13, 5, 6, 7];
const sortedArr = mergeSort(arr);
console.log("Sorted array is:", sortedArr);
